numba/targets/arrayobj.py

```diff
@@ -1739,6 +1783,307 @@ def _increment_indices_array(context, builder, arrty, arr, indices, end_flag=Non
+                builder.store(zero, self.member_ptr)
+            elif logical_dim == self.ndim - 1:
+                # Inside innermost logical dimension, increment index
+                index = builder.load(self.member_ptr)
+                index = increment_index(builder, index)
+                builder.store(index, self.member_ptr)
+
+
+    class TrivialFlatSubIter(BaseSubIter):
+
+        def init_specific(self, context, builder):
+            assert not nditerty.need_shaped_indexing
+
+        def compute_pointer(self, context, builder, indices, arrty, arr):
+            assert len(indices) <= 1, len(indices)
+            return builder.gep(arr.data, indices)
+
+
+    class IndexedSubIter(BaseSubIter):
+
+        def compute_pointer(self, context, builder, indices, arrty, arr):
+            assert len(indices) == self.ndim
+            return cgutils.get_item_pointer(builder, arrty, arr,
+                                            indices, wraparound=False)
+
+
+    class ZeroDimSubIter(BaseSubIter):
+
+        def compute_pointer(self, context, builder, indices, arrty, arr):
+            return arr.data
+
+
+    class ScalarSubIter(BaseSubIter):
+
+        def compute_pointer(self, context, builder, indices, arrty, arr):
+            return arr
+
+
+    class NdIter(cgutils.create_struct_proxy(nditerty)):
+        """
+        .nditer() implementation.
+
+        Note: 'F' layout means the shape is iterated in reverse logical order,
+        so indices and shapes arrays have to be reversed as well.
+        """
+
+        @utils.cached_property
+        def subiters(self):
+            l = []
+            factories = {'flat': FlatSubIter if nditerty.need_shaped_indexing
+                                 else TrivialFlatSubIter,
+                         'indexed': IndexedSubIter,
+                         '0d': ZeroDimSubIter,
+                         'scalar': ScalarSubIter,
+                         }
+            for i, sub in enumerate(nditerty.indexers):
+                kind, start_dim, end_dim, _ = sub
+                member_name = 'index%d' % i
+                factory = factories[kind]
+                l.append(factory(self, member_name, start_dim, end_dim))
+            return l
+
+        def init_specific(self, context, builder, arrtys, arrays):
+            zero = context.get_constant(types.intp, 0)
+
+            # Store inputs
+            self.arrays = context.make_tuple(builder, types.Tuple(arrtys),
+                                             arrays)
+            # Create slots for scalars
+            for i, ty in enumerate(arrtys):
+                if not isinstance(ty, types.Array):
+                    member_name = 'scalar%d' % i
+                    # XXX as_data()?
+                    slot = cgutils.alloca_once_value(builder, arrays[i])
+                    setattr(self, member_name, slot)
+
+            arrays = self._arrays_or_scalars(context, builder, arrtys, arrays)
```

Comment:
I'll take a look, but the naming seems to make sense here. They're all variations on the same thing.